presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonExtract.java

```
@@ -302,11 +306,12 @@ public Slice extract(JsonParser jsonParser)
             if (!jsonParser.hasCurrentToken()) {
                 throw new JsonParseException(jsonParser, "Unexpected end of value");
             }
-
             DynamicSliceOutput dynamicSliceOutput = new DynamicSliceOutput(ESTIMATED_JSON_OUTPUT_SIZE);
-            try (JsonGenerator jsonGenerator = createJsonGenerator(JSON_FACTORY, dynamicSliceOutput)) {
-                jsonGenerator.copyCurrentStructure(jsonParser);
-            }
+            // Write the JSON to output stream with sorted keys
+            SORTED_MAPPER.writeValue((OutputStream) dynamicSliceOutput, SORTED_MAPPER.readValue(jsonParser, Object.class));
+            // nextToken() returns null if the input is parsed correctly,
+            // but will throw an exception if there are trailing characters.
+            jsonParser.nextToken();
```

catch the exception and throw a PrestoException with an invalid_function_argument error code. include the failing jsonInput in the error message.  Also, when there is no exception, test that the value is null (non-null and no exception means there's another valid json object after this one) and throw a PrestoException with invalid_function_argument for that case too.

---

Hi @rschlussel, the condition jsonParser.nextToken() == null is only suitable for json_parse func, but not for json_extract. For example, given the JSON {k1: {...}, k2: v2} and the path $.k1, in this case, the next token would essentially be "FIELD_NAME".
I believe that to validate the entire input, we would need to either implement logic to traverse the JSON structure and verify the validity of all elements or parse the entire input as a JSON object before performing the extraction. However, both of these approaches seem excessive, especially considering that Presto is currently in maintenance mode
Instead, I suggest fixing it to produce Canonicalized Output and throwing a parser error instead of returning null. What do you think?

---

got it. That makes sense.

I'm not sure what you mean "throwing a parser error instead of returning null". Do you mean a parsing error when there are exceptions in the nextToken call you added?  That sounds good to me.